Alerting: Write and Delete multiple alert instances.

joeblubaugh opened the pull request:
Prior to this change, all alert instance writes and deletes happened
individually, in their own database transaction. This change batches up
writes or deletes for a given rule's evaluation loop into a single
transaction before applying it.

Before:
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/services/ngalert/store
BenchmarkAlertInstanceOperations-8           398           2991381 ns/op         1133537 B/op      27703 allocs/op
--- BENCH: BenchmarkAlertInstanceOperations-8
    util.go:127: alert definition: {orgID: 1, UID: FovKXiRVzm} with title: "an alert definition FTvFXmRVkz" interval: 60 created
    util.go:127: alert definition: {orgID: 1, UID: foDFXmRVkm} with title: "an alert definition fovFXmRVkz" interval: 60 created
    util.go:127: alert definition: {orgID: 1, UID: VQvFuigVkm} with title: "an alert definition VwDKXmR4kz" interval: 60 created
PASS
ok      github.com/grafana/grafana/pkg/services/ngalert/store   1.619s
```

After:
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/services/ngalert/store
BenchmarkAlertInstanceOperations-8          1440            816484 ns/op          352297 B/op       6529 allocs/op
--- BENCH: BenchmarkAlertInstanceOperations-8
    util.go:127: alert definition: {orgID: 1, UID: 302r_igVzm} with title: "an alert definition q0h9lmR4zz" interval: 60 created
    util.go:127: alert definition: {orgID: 1, UID: 71hrlmR4km} with title: "an alert definition nJ29_mR4zz" interval: 60 created
    util.go:127: alert definition: {orgID: 1, UID: Cahr_mR4zm} with title: "an alert definition ja2rlmg4zz" interval: 60 created
PASS
ok      github.com/grafana/grafana/pkg/services/ngalert/store   1.383s
```

So we cut time by about 75% and memory allocations by about 60% when
storing and deleting 100 instances.

joeblubaugh committed:
Alerting: Write and Delete multiple alert instances.

Priod to this change, all alert instance writes and deletes happened
individually, in their own database transaction. This change batches up
writes or deletes for a given rule's evaluation loop into a single
transaction before applying it.

Before:
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/services/ngalert/store
BenchmarkAlertInstanceOperations-8           398           2991381 ns/op         1133537 B/op      27703 allocs/op
--- BENCH: BenchmarkAlertInstanceOperations-8
    util.go:127: alert definition: {orgID: 1, UID: FovKXiRVzm} with title: "an alert definition FTvFXmRVkz" interval: 60 created
    util.go:127: alert definition: {orgID: 1, UID: foDFXmRVkm} with title: "an alert definition fovFXmRVkz" interval: 60 created
    util.go:127: alert definition: {orgID: 1, UID: VQvFuigVkm} with title: "an alert definition VwDKXmR4kz" interval: 60 created
PASS
ok      github.com/grafana/grafana/pkg/services/ngalert/store   1.619s
```

After:
```
goos: darwin
goarch: arm64
pkg: github.com/grafana/grafana/pkg/services/ngalert/store
BenchmarkAlertInstanceOperations-8          1440            816484 ns/op          352297 B/op       6529 allocs/op
--- BENCH: BenchmarkAlertInstanceOperations-8
    util.go:127: alert definition: {orgID: 1, UID: 302r_igVzm} with title: "an alert definition q0h9lmR4zz" interval: 60 created
    util.go:127: alert definition: {orgID: 1, UID: 71hrlmR4km} with title: "an alert definition nJ29_mR4zz" interval: 60 created
    util.go:127: alert definition: {orgID: 1, UID: Cahr_mR4zm} with title: "an alert definition ja2rlmg4zz" interval: 60 created
PASS
ok      github.com/grafana/grafana/pkg/services/ngalert/store   1.383s
```

So we cut time by about 75% and memory allocations by about 60% when
storing and deleting 100 instances.

joeblubaugh commented on pkg/services/ngalert/models/instance.go:
```diff
@@ -44,8 +44,12 @@ func (i InstanceStateType) IsValid() bool {
+
 // SaveAlertInstanceCommand is the query for saving a new alert instance.
-type SaveAlertInstanceCommand struct {
+type SaveAlertInstanceCommandFields struct {
```
I need a better name for this, but I'm not sure what.

joeblubaugh replied:
Eliminated this struct.

joeblubaugh commented on pkg/services/ngalert/store/instance_database.go:
```diff
@@ -12,9 +12,15 @@ import (
+	DeleteAlertInstances(ctx context.Context, keys ...InstanceKey) error
+}
+
+type InstanceKey struct {
```
Should I put this in `models`?

joeblubaugh replied:
Yes

joeblubaugh commented on pkg/services/ngalert/store/instance_database.go:
```diff
@@ -149,11 +177,18 @@ func (st DBstore) FetchOrgIds(ctx context.Context) ([]int64, error) {
-		_, err := sess.Exec("DELETE FROM alert_instance WHERE rule_org_id = ? AND rule_uid = ? AND labels_hash = ?", orgID, ruleUID, labelsHash)
-		if err != nil {
-			return err
+		for _, k := range keys {
```
TODO: should probably use a row limit here as well, so our transactions don't get too big.

grobinson-grafana commented on pkg/services/ngalert/models/instance.go:
```diff
@@ -44,8 +44,12 @@ func (i InstanceStateType) IsValid() bool {
+
 // SaveAlertInstanceCommand is the query for saving a new alert instance.
-type SaveAlertInstanceCommand struct {
+type SaveAlertInstanceCommandFields struct {
```
```suggestion
type SaveAlertInstanceCommand struct {
```

grobinson-grafana replied:
I think it makes sense to keep this struct called `SaveAlertInstanceCommand` and then the other struct either `SaveAlertInstancesBulkCommand` or `SaveAlertInstancesBatchCommand`?

joeblubaugh replied:
I ended up making the function variadic, so this is moot.

grobinson-grafana commented on pkg/services/ngalert/store/instance_database.go:
```diff
@@ -87,43 +93,65 @@ func (st DBstore) ListAlertInstances(ctx context.Context, cmd *models.ListAlertI
+				limit = len(cmd.Instances)
+			}
+
+			for _, fields := range cmd.Instances[i:limit] {
```
No error if we have exceeded the maximum number of rows to update?

joeblubaugh replied:
No, we just move on and save the next chunk.

grobinson-grafana commented on pkg/services/ngalert/store/instance_database.go:
```diff
@@ -87,43 +93,65 @@ func (st DBstore) ListAlertInstances(ctx context.Context, cmd *models.ListAlertI
-			[]string{"rule_org_id", "rule_uid", "labels", "labels_hash", "current_state", "current_reason", "current_state_since", "current_state_end", "last_eval_time"})
-		_, err = sess.SQL(upsertSQL, params...).Query()
+// SaveAlertInstances saves all the provided alert instances in a single write transaction.
+func (st DBstore) SaveAlertInstances(ctx context.Context, cmd *models.SaveAlertInstancesCommand) error {
```
Perhaps we should validate the cmd before opening the tx? For example, the number of rows must be less than **N**?

joeblubaugh commented on pkg/services/ngalert/store/instance_database.go:
```diff
@@ -87,43 +93,65 @@ func (st DBstore) ListAlertInstances(ctx context.Context, cmd *models.ListAlertI
+// SaveAlertInstances saves all the provided alert instances in a single write transaction.
+func (st DBstore) SaveAlertInstances(ctx context.Context, cmd *models.SaveAlertInstancesCommand) error {
+	rowLimit := 400
+	for i := 0; i < len(cmd.Instances); i += rowLimit {
```
We're processing the write in increments of 400 rows. 400 is chosen largely based on "intuition". I could benchmark against SQLite and see when things get bad.

joeblubaugh committed:
Make SaveAlertInstances variadic, eliminate redundant types

joeblubaugh committed:
Remove unnecessary pointer argument

yuri-tceretian commented on pkg/services/ngalert/models/instance.go:
```diff
@@ -78,11 +70,7 @@ type ListAlertInstancesQuery struct {
-		return fmt.Errorf("alert instance is invalid because it is nil")
-	}
-
+func ValidateAlertInstance(alertInstance AlertInstance) error {
```
do we really need validation for this struct? the only consumer and producer of this data - state manager.

joeblubaugh replied:
I haven't changed this function, save changing the type from a struct pointer to a struct. Is this blocking a merge, or do you have other feedback?

grobinson-grafana commented:
I would like to do a quick review before we merge if that's OK!

grobinson-grafana commented on pkg/services/ngalert/models/instance_labels.go:
```diff
@@ -42,7 +42,7 @@ func (il *InstanceLabels) StringKey() (string, error) {
 	b, err := json.Marshal(tl)
 	if err != nil {
-		return "", fmt.Errorf("can not gereate key due to failure to encode labels: %w", err)
+		return "", fmt.Errorf("can not generate key due to failure to encode labels: %w", err)
```
```suggestion
		return "", fmt.Errorf("could not generate key due to failure to encode labels: %w", err)
```

joeblubaugh committed:
Fix bugs with bulk writes and deletes.

We need to lower the row limit in upserts based on the # of columns
being read or written.

joeblubaugh committed:
Bind to a prepared statement instead of building big statements.

No performance difference in million-row tests.

joeblubaugh commented on pkg/services/ngalert/state/manager.go:
```diff
@@ -370,22 +368,27 @@ func (st *Manager) annotateState(ctx context.Context, alertRule *ngModels.AlertR
-		_, ok := states[s.CacheId]
-		if !ok && isItStale(evaluatedAt, s.LastEvaluationTime, alertRule.IntervalSeconds) {
+		// Is the cached state in our recently processed results? If not, is it stale?
+		if _, ok := newStates[s.CacheId]; !ok && isItStale(evaluatedAt, s.LastEvaluationTime, alertRule.IntervalSeconds) {
```
“newStates” are the states generated by a single rule evaluation. We’re looking for states for the same rule that are missing from the most recent evaluation. I’m very open to a new name for this variable because I got confused when editing this, too.

joeblubaugh commented on pkg/services/ngalert/store/instance_database.go:
```diff
@@ -149,12 +162,34 @@ func (st DBstore) FetchOrgIds(ctx context.Context) ([]int64, error) {
+// DeleteAlertInstances deletes instances with the provided keys. It
+// writes transactions with a maximum count of 400 rows. Larger writes are
+// split into multiple transactions.
+func (st DBstore) DeleteAlertInstances(ctx context.Context, keys ...models.AlertInstanceKey) error {
```
My latest change makes these both single transactions with many statements, instead of big concatenation statements. Benchmarking the two didn’t show much performance difference, and it lets us use a single transaction like we ought to.

joeblubaugh committed:
Merge branch 'main' into joe/save-multiple-instances

joeblubaugh committed:
Merge main into joe/save-multiple-instances

joeblubaugh commented:
@grobinson-grafana this thing is going to get kind of weird due to how long the PR hung out. I'm rebasing on main before submitting again. That will involve a force push.

joeblubaugh commented:
See #54072
